// socks/docs/examples/javascript/lib/AsyncParallelBailHook.js
: DateExpression | {
        /**
         * The date to which the operator is applied. <dateExpression> must be a valid expression that resolves to a Date, a Timestamp, or an ObjectID.
         */
        date: DateExpression;
        /**
         * The timezone of the operation result. <tzExpression> must be a valid expression that resolves to a string formatted as either an Olson Timezone Identifier or a UTC Offset. If no timezone is provided, the result is displayed in UTC.
         */
        timezone?: tzExpression;
      };
    }

    export interface Year {
      /**
       * Returns the year for a date as a number (e.g. 2014).
       *
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/year/#mongodb-expression-exp.-year
       */
      $year: DateExpression | {
        /**
         * The date to which the operator is applied. <dateExpression> must be a valid expression that resolves to a Date, a Timestamp, or an ObjectID.
         */
        date: DateExpression;
        /**
         * The timezone of the operation result. <tzExpression> must be a valid expression that resolves to a string formatted as either an Olson Timezone Identifier or a UTC Offset. If no timezone is provided, the result is displayed in UTC.
         */
        timezone?: tzExpression;
      };
    }

    export interface And {
      /**
       * Returns true only when all its expressions evaluate to true. Accepts any number of argument expressions.
       *
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/and/#mongodb-expression-exp.-and
       */
      $and: (Expression | Record<string, Expression>)[];
    }

    export interface Not {
      /**
       * Returns the boolean value that is the opposite of its argument expression. Accepts a single argument expression.
       *
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/not/#mongodb-expression-exp.-not
       */
      $not: [Expression];
    }

    export interface Or {
      /**
       * Returns true when any of its expressions evaluates to true. Accepts any number of argument expressions.
       *
       * @see https://www.mongodb.com/docs/manual/reference/operator/agg
