// socks/docs/dist/ExplorerBase.js
second, and -1 if the first string is less than the second.
       *
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/strcasecmp/#mongodb-expression-exp.-strcasecmp
       */
      $strcasecmp: [StringExpression, StringExpression];
    }

    export interface Substr {
      /**
       * Deprecated. Use $substrBytes or $substrCP.
       *
       * @deprecated 3.4
       * @alias {Expression.SubstrBytes}
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/substr/#mongodb-expression-exp.-substr
       */
      $substr: [StringExpression, number, number];
    }

    export interface SubstrBytes {
      /**
       * Returns the substring of a string. Starts with the character at the specified UTF-8 byte index (zero-based) in the string and continues for the specified number of bytes.
       *
       * @version 3.4
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/substrBytes/#mongodb-expression-exp.-substrBytes
       */
      $substrBytes: [StringExpression, number, number];
    }

    export interface SubstrCP {
      /**
       * Returns the substring of a string. Starts with the character at the specified UTF-8 code point (CP) index (zero-based) in the string and continues for the number of code points specified.
       *
       * @version 3.4
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/substrCP/#mongodb-expression-exp.-substrCP
       */
      $substrCP: [StringExpression, number, number];
    }

    export interface ToLower {
      /**
       * Converts a string to lowercase. Accepts a single argument expression.
       *
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/toLower/#mongodb-expression-exp.-toLower
       */
      $toLower: StringExpression;
    }

    export interface ToString {
      /**
       * Converts value to a string.
       *
       * @version 4.0
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/toString/#mongodb-expression-exp.-toString
       */
      $toString: Expression;
    }

    export interface Trim {
      /**
       * Removes whitespace or the specified characters from the beginning and end of a string.
       *
       * @version 4.0
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/trim/#mongodb-expression-exp.-trim
       */
      $trim: {
        /**
         * The string to trim. The argument can be any valid expression that resolves to a string. For more information on expressions, see Expressions.
         */
        input: StringExpression;
        /**
         * The character(s) to trim from the beginning of the input.
         *
         * The argument can be any valid expression that resolves to a string. The $trim operator breaks down the string into individual UTF code point to trim from input.
         *
         * If unspecified, $trim removes whitespace characters, including the null character. For the list of whitespace characters, see Whitespace Characters.
         */
        chars?: StringExpression;
      };
    }

    export interface ToUpper {
      /**
       * Converts a string to uppercase. Accepts a single argument expression.
       *
       * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/toUpper/#mongodb-expression-exp.-toUpper
       */
      $toUpper: StringExpression;
